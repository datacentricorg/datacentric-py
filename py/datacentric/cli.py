# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
from enum import IntEnum

from datacentric.storage.class_info import ClassInfo
from datacentric.storage.env_type import EnvType
from datacentric.storage.context import Context
from datacentric.storage.data import Data
from datacentric.storage.mongo.mongo_server import MongoServer
from datacentric.storage.mongo.temporal_mongo_data_source import TemporalMongoDataSource
from datacentric.schema.declaration.type_decl import TypeDecl
from datacentric.schema.declaration.enum_decl import EnumDecl
from datacentric.schema.generator import to_type_declaration, to_enum_declaration

parser = argparse.ArgumentParser()
sub_parsers = parser.add_subparsers(help='commands', dest='command')

# Schema command
schema_parser = sub_parsers.add_parser('schema', help='Write schema to data source.')
schema_parser.add_argument('--packages', '-p', nargs='+', required=True, help='Generate schema from provided packages')
# Change short option to avoid -h conflict. Users always expect -h to work
schema_parser.add_argument('--host', '-o', type=str, required=False,
                           help='Db Host. Fallbacks to standard if not provided')
schema_parser.add_argument('--env', '-e', type=str, required=True, help='Environment type')
schema_parser.add_argument('--group', '-g', type=str, required=True, help='Environment group')
schema_parser.add_argument('--name', '-n', type=str, required=True, help='Environment name')

# Run command
run_parser = sub_parsers.add_parser('run', help='Execute handler')
run_parser.add_argument('--source', '-s', required=True,
                        help='Source environment - folder for file storage and connection string for DB.')
run_parser.add_argument('--env', '-e', type=str, required=True, help='Environment type')
run_parser.add_argument('--group', '-g', type=str, required=True, help='Environment group')
run_parser.add_argument('--name', '-n', type=str, required=True, help='Environment name')
run_parser.add_argument('--dataset', '-d', required=True, help='Setting specifies data set name.')
run_parser.add_argument('--key', '-k', required=True, help='Key of entity.')
run_parser.add_argument('--type', '-t', required=True, help='Type handler belongs.')
# Change short option to avoid -h conflict. Users always expect -h to work
run_parser.add_argument('--handler', '-l', required=True, help='Handler name to execute.')
run_parser.add_argument('--arguments', '-a', help='Space separated handler arguments in name=value format.')


def schema(args):
    context = Context()

    data_source = TemporalMongoDataSource()
    data_source.env_type = EnvType[args.env]
    data_source.env_group = args.group
    data_source.env_name = args.name
    if args.host is not None:
        data_source.mongo_server = MongoServer.create_key(mongo_server_uri=args.host)
    data_source.init(context)

    common = data_source.get_common_or_none()
    if common is None:
        common = data_source.create_common()

    context.data_source = data_source
    context.data_set = common

    datas = ClassInfo.get_derived_types('datacentric', Data)
    enums = ClassInfo.get_derived_types('datacentric', IntEnum)
    type_declarations = [to_type_declaration(x) for x in datas]
    enum_declarations = [to_enum_declaration(x) for x in enums]

    # Save declarations to db
    context.data_source.save_many(TypeDecl, type_declarations, context.data_set)
    context.data_source.save_many(EnumDecl, enum_declarations, context.data_set)


def run(args):
    pass


if __name__ == '__main__':
    res = parser.parse_args()
    command = res.command
    if command == 'schema':
        schema(res)
    elif command == 'run':
        run(res)
    else:
        print(f'Unknown command: {command}.')
        parser.print_help()
