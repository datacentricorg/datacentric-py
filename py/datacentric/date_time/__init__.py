from datacentric.date_time.iso_day_of_week import IsoDayOfWeek
from datacentric.date_time.local_minute import LocalMinute
