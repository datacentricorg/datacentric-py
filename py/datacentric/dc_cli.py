import argparse
from enum import IntEnum

from datacentric.storage.class_info import ClassInfo
from datacentric.storage.env_type import EnvType
from datacentric.storage.context import Context
from datacentric.storage.data import Data
from datacentric.storage.mongo.mongo_server import MongoServer
from datacentric.storage.mongo.temporal_mongo_data_source import TemporalMongoDataSource
from datacentric.schema.declaration.type_decl import TypeDecl
from datacentric.schema.declaration.enum_decl import EnumDecl
from datacentric.schema.generator import to_type_declaration, to_enum_declaration

parser = argparse.ArgumentParser()
sub_parsers = parser.add_subparsers(help='commands', dest='command')

# Schema command
schema_parser = sub_parsers.add_parser('schema', help='Write schema to data source.')
schema_parser.add_argument('--packages', '-p', nargs='+', required=True, help='Generate schema from provided packages')
# Change short option to avoid -h conflict. Users always expect -h to work
schema_parser.add_argument('--host', '-o', type=str, required=False,
                           help='Db Host. Fallbacks to standard if not provided')
schema_parser.add_argument('--env', '-e', type=str, required=True, help='Environment type')
schema_parser.add_argument('--group', '-g', type=str, required=True, help='Environment group')
schema_parser.add_argument('--name', '-n', type=str, required=True, help='Environment name')


def schema(args):
    context = Context()

    data_source = TemporalMongoDataSource()
    data_source.env_type = EnvType[args.env]
    data_source.env_group = args.group
    data_source.env_name = args.name
    if args.host is not None:
        data_source.mongo_server = MongoServer.create_key(mongo_server_uri=args.host)
    data_source.init(context)

    common = data_source.get_common_or_none()
    if common is None:
        common = data_source.create_common()

    context.data_source = data_source
    context.data_set = common

    datas = ClassInfo.get_derived_types('datacentric', Data)
    enums = ClassInfo.get_derived_types('datacentric', IntEnum)
    type_declarations = [to_type_declaration(x) for x in datas]
    enum_declarations = [to_enum_declaration(x) for x in enums]

    # Save declarations to db
    context.data_source.save_many(TypeDecl, type_declarations, context.data_set)
    context.data_source.save_many(EnumDecl, enum_declarations, context.data_set)


if __name__ == '__main__':
    res = parser.parse_args()
    if res.command == 'schema':
        schema(res)
