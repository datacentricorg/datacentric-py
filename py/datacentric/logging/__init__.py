from datacentric.logging.in_memory_log import InMemoryLog
from datacentric.logging.log_entry_type import LogEntryType
