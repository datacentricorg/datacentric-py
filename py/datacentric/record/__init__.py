from datacentric.record.data import Data
from datacentric.record.key import Key

from datacentric.record.record import Record
from datacentric.record.typed_record import TypedRecord
from datacentric.record.root_record import RootRecord
from datacentric.record.deleted_record import DeletedRecord

from datacentric.record.typed_key import TypedKey
