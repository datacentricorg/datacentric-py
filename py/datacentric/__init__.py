name = "datacentric"

from datacentric.file_system.text_writer import TextWriter
from datacentric.file_system.console_writer import ConsoleWriter
from datacentric.file_system.string_writer import StringWriter
from datacentric.file_system.file_writer import FileWriter
from datacentric.date_time.local_date import LocalDate
from datacentric.date_time.local_time import LocalTime
from datacentric.date_time.local_date_time import LocalDateTime
from datacentric.date_time.instant import Instant
from datacentric.date_time.iso_day_of_week import IsoDayOfWeek
from datacentric.storage.env_type import EnvType
from datacentric.storage.context import Context
from datacentric.storage.key import Key
from datacentric.storage.record import Record
from datacentric.storage.key_util import KeyUtil
from datacentric.storage.data_source import DataSource, DataSourceKey
from datacentric.storage.mongo.mongo_data_source import MongoDataSource
from datacentric.storage.mongo.temporal_mongo_data_source import TemporalMongoDataSource
from datacentric.testing.unit_test import UnitTestKey, UnitTest
from datacentric.date_time.zone import ZoneKey, Zone
from datacentric.storage.mongo.temporal_mongo_unit_test_context import TemporalMongoUnitTestContext
from datacentric.storage.unit_test_context import UnitTestContext
from datacentric.testing.unit_test_key import UnitTestKey
from datacentric.testing.unit_test_complexity import UnitTestComplexity
from datacentric.schema.declaration.element_decl import ElementDecl
from datacentric.schema.declaration.enum_decl import EnumDecl, EnumDeclKey
from datacentric.schema.declaration.enum_item import EnumItem
from datacentric.schema.declaration.handler_declare_block import HandlerDeclareBlock
from datacentric.schema.declaration.handler_declare_item import HandlerDeclareItem
from datacentric.schema.declaration.handler_implement_block import HandlerImplementBlock
from datacentric.schema.declaration.handler_implement_item import HandlerImplementItem
from datacentric.schema.declaration.index_element import IndexElement
from datacentric.schema.declaration.index_elements import IndexElements
from datacentric.schema.declaration.language import Language
from datacentric.schema.declaration.module import Module, ModuleKey
from datacentric.schema.declaration.param_decl import ParamDecl
from datacentric.schema.declaration.type_decl import TypeDecl, TypeDeclKey
