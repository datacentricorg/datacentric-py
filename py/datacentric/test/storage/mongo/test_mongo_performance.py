#  Copyright (C) 2013-present The DataCentric Authors.
#  #
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  #
#     http://www.apache.org/licenses/LICENSE-2.0
#  #
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

import unittest

from pymongo.database import Database
from pymongo import DESCENDING

from datacentric import TemporalMongoUnitTestContext, Context


class TestMongoPerformance(unittest.TestCase):
    RECORD_COUNT = 10  # 300_000
    DATASET_COUNT = 2  # 10
    VERSION_COUNT = 2  # 10
    ARRAY_SIZE = 10  # 1_000

    def get_db(self, context: Context) -> Database:
        return context.data_source.db

    def insert_records_a(self, context: Context):
        db = self.get_db(context)
        records = []
        for i in range(TestMongoPerformance.RECORD_COUNT):
            rec = {'_id': 'KeyPrefix' + str(i),
                   'DoubleElement': float(i),
                   'IntElement': i,
                   'VersionElement': 0}
            if TestMongoPerformance.ARRAY_SIZE > 0:
                rec['ArrayElement'] = [float(x) for x in range(TestMongoPerformance.ARRAY_SIZE)]
            records.append(rec)

        # Unique index on _id is created automatically
        collection = db.get_collection('A')
        collection.create_index([('DoubleElement', DESCENDING)])
        collection.insert_many(records)
        context.log.verify(f'Inserted {TestMongoPerformance.RECORD_COUNT} records.')

    def test_insert_a(self):
        with TemporalMongoUnitTestContext() as context:
            self.insert_records_a(context)
