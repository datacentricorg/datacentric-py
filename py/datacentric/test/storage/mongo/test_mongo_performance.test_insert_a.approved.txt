Verify: Inserted 10 records.
Verify: Test completed successfully.
